homeassistant/helpers/location.py

```diff
@@ -41,3 +48,61 @@ def closest(
             longitude,
         ),
     )
+
+
+def coordinates(
+    hass: HomeAssistantType, entity_id: str, recursion_history: Optional[list] = None
+) -> Optional[str]:
+    """Get the location from the entity state or attributes."""
+    entity = hass.states.get(entity_id)
+
+    if entity is None:
+        _LOGGER.error("Unable to find entity %s", entity_id)
+        return None
+
+    # Check if the entity has location attributes
+    if has_location(entity):
+        return _get_location_from_attributes(entity)
+
+    # Check if device is in a zone
+    zone_entity = hass.states.get(f"zone.{entity.state}")
+    if has_location(zone_entity):  # type: ignore
+        _LOGGER.debug(
+            "%s is in %s, getting zone location", entity_id, zone_entity.entity_id  # type: ignore
+        )
+        return _get_location_from_attributes(zone_entity)  # type: ignore
+
+    # Resolve nested entity
+    if recursion_history is None:
+        recursion_history = []
+    recursion_history.append(entity_id)
+    if entity.state in recursion_history:
+        _LOGGER.error(
+            "Circular Reference detected. The state of %s has already been checked.",
+            entity.state,
+        )
+        return None
+    _LOGGER.debug("Getting nested entity for state: %s", entity.state)
+    nested_entity = hass.states.get(entity.state)
+    if nested_entity is not None:
+        _LOGGER.debug("Resolving nested entity_id: %s", entity.state)
+        return coordinates(hass, entity.state, recursion_history)
+
+    # Check if state is valid coordinate set
+    try:
+        cv.gps(entity.state.split(","))
+    except vol.Invalid:
+        _LOGGER.error(
+            "The state of %s is not a valid set of coordinates: %s",
```

Comment:
Is this the right error? So many other things could have gone wrong for us to end up here.

Reply:
To end up here the function checked:

1. The entity is present in the state machine
1. The entity is not a device_tracker
1. The entity has a state but it is not the entity_id on another entity
1. The entity has a state but it is not a properly formatted set of gps coordinates.

According to this function it got passed a wrong entity_id.

Reply:
Wouldn't this make more sense: Entity %s does not contain a location and does not point at an entity that does

Reply:
Makes much more sense. I pushed the change.